File: java/exercises/practice/unit_test_fd6952da_dcf4_4d9f_a884_8f61b43fa445/task_schedule/src/test/java/TaskSchedulerTest.java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.util.*;
import task_schedule.Task;
import task_schedule.TaskScheduler;

public class TaskSchedulerTest {

    @Test
    public void testSingleTask() {
        List<Task> tasks = new ArrayList<>();
        // Task(id, duration, deadline, dependencies)
        tasks.add(new Task(0, 5, 10, Collections.emptyList()));
        TaskScheduler scheduler = new TaskScheduler();
        int result = scheduler.minimizeMaximumLateness(tasks);
        // A single task can always be delayed so that it finishes exactly at its deadline.
        assertEquals(0, result);
    }

    @Test
    public void testLinearDependencies() {
        List<Task> tasks = new ArrayList<>();
        // Two tasks in a chain: Task 1 depends on Task 0.
        tasks.add(new Task(0, 3, 5, Collections.emptyList()));
        tasks.add(new Task(1, 4, 11, Arrays.asList(0)));
        TaskScheduler scheduler = new TaskScheduler();
        int result = scheduler.minimizeMaximumLateness(tasks);
        // It is possible to schedule these tasks so that each finishes by its deadline.
        assertEquals(0, result);
    }

    @Test
    public void testInsufficientSlack() {
        List<Task> tasks = new ArrayList<>();
        // A dependency chain where deadlines are too tight.
        tasks.add(new Task(0, 5, 6, Collections.emptyList()));
        tasks.add(new Task(1, 4, 8, Arrays.asList(0)));
        TaskScheduler scheduler = new TaskScheduler();
        int result = scheduler.minimizeMaximumLateness(tasks);
        // Even with optimal delays, the dependent task will finish 2 time units late.
        assertEquals(2, result);
    }

    @Test
    public void testParallelTasks() {
        List<Task> tasks = new ArrayList<>();
        // Two independent tasks that can be executed in parallel.
        tasks.add(new Task(0, 4, 5, Collections.emptyList()));
        tasks.add(new Task(1, 6, 8, Collections.emptyList()));
        TaskScheduler scheduler = new TaskScheduler();
        int result = scheduler.minimizeMaximumLateness(tasks);
        // Both tasks can be delayed to finish exactly at their deadlines.
        assertEquals(0, result);
    }

    @Test
    public void testComplexGraph() {
        List<Task> tasks = new ArrayList<>();
        // A more intricate dependency graph.
        tasks.add(new Task(0, 5, 10, Collections.emptyList()));
        tasks.add(new Task(1, 3, 15, Arrays.asList(0)));
        tasks.add(new Task(2, 7, 20, Arrays.asList(1)));
        tasks.add(new Task(3, 2, 12, Arrays.asList(0)));
        TaskScheduler scheduler = new TaskScheduler();
        int result = scheduler.minimizeMaximumLateness(tasks);
        // With optimal scheduling, tasks can meet their deadlines.
        assertEquals(0, result);
    }

    @Test
    public void testImpossibleSchedule() {
        List<Task> tasks = new ArrayList<>();
        // A task whose duration is longer than its deadline makes scheduling impossible.
        tasks.add(new Task(0, 20, 10, Collections.emptyList()));
        TaskScheduler scheduler = new TaskScheduler();
        int result = scheduler.minimizeMaximumLateness(tasks);
        // Indicate failure since it is impossible to finish before the deadline.
        assertEquals(Integer.MAX_VALUE, result);
    }

    @Test
    public void testMultipleDependencies() {
        List<Task> tasks = new ArrayList<>();
        // A task with multiple dependencies.
        tasks.add(new Task(0, 3, 10, Collections.emptyList()));
        tasks.add(new Task(1, 2, 8, Collections.emptyList()));
        tasks.add(new Task(2, 4, 15, Arrays.asList(0, 1)));
        tasks.add(new Task(3, 1, 12, Arrays.asList(0)));
        tasks.add(new Task(4, 2, 20, Arrays.asList(2, 3)));
        TaskScheduler scheduler = new TaskScheduler();
        int result = scheduler.minimizeMaximumLateness(tasks);
        // Optimal scheduling should allow all tasks to meet their deadlines.
        assertEquals(0, result);
    }
}

File: cpp/exercises/practice/unit_test_43f5c9d6_6e2b_4b47_baf1_ccf8ff00be87/task_scheduler/task_scheduler_test.cpp
#include "catch.hpp"
#include "task_scheduler.h"

#include <vector>

using std::vector;

TEST_CASE("single_task_feasible") {
    // One task with no dependencies, and deadline is met.
    // Expected maximum lateness is 0.
    vector<Task> tasks = {
        {0, 5, 10, {}}
    };
    int result = schedule_tasks(1, 1, tasks);
    REQUIRE(result == 0);
}

TEST_CASE("chain_tasks_feasible") {
    // Three tasks in a dependency chain: 0 -> 1 -> 2
    // Their deadlines are set exactly to the cumulative finish times.
    vector<Task> tasks = {
        {0, 4, 4, {}},
        {1, 4, 8, {0}},
        {2, 4, 12, {1}}
    };
    int result = schedule_tasks(3, 1, tasks);
    REQUIRE(result == 0);
}

TEST_CASE("parallel_tasks_feasible") {
    // Four tasks with a mix of independent tasks and dependencies.
    // Task 1 depends on task 0; task 2 depends on tasks 0 and 1.
    // With two machines, an optimal schedule exists with zero lateness.
    vector<Task> tasks = {
        {0, 5, 10, {}},
        {1, 3, 12, {0}},
        {2, 4, 15, {0, 1}},
        {3, 2, 20, {}}
    };
    int result = schedule_tasks(4, 2, tasks);
    REQUIRE(result == 0);
}

TEST_CASE("single_task_impossible") {
    // A single task that cannot meet its deadline even if scheduled immediately.
    vector<Task> tasks = {
        {0, 10, 5, {}}
    };
    int result = schedule_tasks(1, 1, tasks);
    REQUIRE(result == -1);
}

TEST_CASE("dependent_task_impossible") {
    // Two tasks in a dependency chain where the first task cannot finish by its deadline.
    // Even with optimal scheduling on one machine, it's impossible to meet all deadlines.
    vector<Task> tasks = {
        {0, 5, 3, {}},
        {1, 5, 15, {0}}
    };
    int result = schedule_tasks(2, 1, tasks);
    REQUIRE(result == -1);
}

TEST_CASE("complex_dag_feasible") {
    // Six tasks forming a complex DAG (dependencies form a tree structure).
    // With two machines, a schedule exists that meets all deadlines.
    // Task structure:
    //   Task 0: no dependencies.
    //   Task 1 and Task 2: depend on Task 0.
    //   Task 3: depends on Task 1.
    //   Task 4: depends on Task 2.
    //   Task 5: depends on both Task 3 and Task 4.
    vector<Task> tasks = {
        {0, 3, 5, {}},
        {1, 2, 6, {0}},
        {2, 4, 10, {0}},
        {3, 3, 12, {1}},
        {4, 2, 12, {2}},
        {5, 1, 15, {3, 4}}
    };
    int result = schedule_tasks(6, 2, tasks);
    REQUIRE(result == 0);
}